solr/core/src/java/org/apache/solr/core/TrackingDirectoryReader.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.core;
+
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.LongAdder;
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.index.DirectoryReader;
+import org.apache.lucene.index.FilterDirectoryReader;
+import org.apache.lucene.index.FilterLeafReader;
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.NumericDocValues;
+import org.apache.lucene.index.PointValues;
+import org.apache.lucene.index.SortedDocValues;
+import org.apache.lucene.index.SortedNumericDocValues;
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.index.Terms;
+import org.apache.lucene.index.VectorValues;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TrackingDirectoryReader extends FilterDirectoryReader {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  private static final Map<String, LongAdder> fieldUsage = new ConcurrentHashMap<>();
+
+  public TrackingDirectoryReader(
+      DirectoryReader in, TrackingSubReaderWrapper trackingSubReaderWrapper) throws IOException {
+    super(in, trackingSubReaderWrapper);
+  }
+
+  public static DirectoryReader wrap(DirectoryReader in) throws IOException {
+    return new TrackingDirectoryReader(in, new TrackingSubReaderWrapper());
+  }
+
+  @Override
+  protected DirectoryReader doWrapDirectoryReader(DirectoryReader in) throws IOException {
+    return wrap(in);
+  }
+
+  @Override
+  public CacheHelper getReaderCacheHelper() {
+    return in.getReaderCacheHelper();
+  }
+
+  private static void trackUsage(String field, String type) {
+    String key = field + '_' + type;
```

Comment:
won't it blow up heap? Couldn't be better to use a kind of tuple as a key, or just String.intern() ?

Reply:
Not sure its going to matter? very short lived object creation that will be immediately gc'd for the key? Not sure how a tuple would be better?

Reply:
I pushed e3e2364 which uses `AbstractMap.SimpleImmutableEntry` but the heap usage here shouldn't be in `trackUsage`

its going to be in the fieldUsage map if there are tons and tons of fields. each field would be a long entry plus the field name.